src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,16 +417,365 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+        {
+            ODataDeserializerContext nestedReadContext = null;
+            Routing.ODataPath path = readContext.Path;
+
+            try
+            {
+                // this code attempts to make sure that the path is always correct for the level that we are reading.
+                if (edmProperty == null)
+                {
+                    ODataNestedResourceInfo nestedInfo = resourceInfoWrapper.NestedResourceInfo;
+
+                    IEdmType segmentType = null;
+                    string propertyTypeName = nestedInfo.TypeAnnotation?.TypeName;
+                    if (!string.IsNullOrEmpty(propertyTypeName))
+                    {
+                        segmentType = readContext.Model.FindType(propertyTypeName);
+                    }
+
+                    // could it be a problem later that the navigationSource is null?
+                    DynamicPathSegment pathSegment = new DynamicPathSegment(
+                       nestedInfo.Name,
+                       segmentType,
+                       null,
+                       nestedInfo.IsCollection != true
+                       );
+
+                    path = AppendToPath(path, pathSegment);
+                }
+                else
+                {
+                    if (edmProperty.PropertyKind == EdmPropertyKind.Navigation)
+                    {
+                        Contract.Assert(readContext.Path.NavigationSource != null, "Navigation property segment with null navigationSource");
+                        IEdmNavigationProperty navigationProperty = edmProperty as IEdmNavigationProperty;
+                        IEdmNavigationSource parentNavigationSource = readContext.Path.NavigationSource;
+                        IEdmNavigationSource navigationSource = parentNavigationSource.FindNavigationTarget(navigationProperty);
+
+                        if (navigationProperty.ContainsTarget)
+                        {
+                            path = AppendToPath(path, new NavigationPropertySegment(navigationProperty, navigationSource), navigationProperty.DeclaringType, parentNavigationSource);
+                        }
+                        else
+                        {
+                            path = new Routing.ODataPath(new ODataUriParser(readContext.Model, new Uri(navigationSource.Path.Path, UriKind.Relative)).ParsePath());
+                        }
+                    }
+                    else
+                    {
+                        IEdmStructuralProperty structuralProperty = edmProperty as IEdmStructuralProperty;
+                        path = AppendToPath(path, new PropertySegment(structuralProperty), structuralProperty.DeclaringType, null);
+                    }
+                }
+
+                nestedReadContext = BuildNestedContextFromCurrentContext(readContext, path);
+            }
+            catch
```

Comment:
We should catch the exception that we should anticipate. Otherwise this is too general. As I was testing, a `NullReferenceException` thrown from `AppendToPath` got caught here. We should check for nulls instead of relying on this catch all exception block
![image](https://user-images.githubusercontent.com/28291332/184600954-7608806d-a511-48a0-96fc-b06ef6f8e6a6.png)